telluric/georaster.py

```diff
@@ -101,21 +103,97 @@ def merge_all(rasters, roi=None, dest_resolution=None, merge_strategy=MergeStrat
-    projected_rasters = [raster for raster in projected_rasters if raster is not None]
-    func = partial(_merge, merge_strategy=merge_strategy)  # type: Callable[[_Raster, _Raster], _Raster]
-    raster = reduce(func, projected_rasters, empty)
-    return empty.copy_with(image=raster.image, band_names=raster.band_names)
+    # Create a list of single band rasters
+    all_band_names, projected_rasters = _prepare_rasters(rasters, merge_strategy, empty)
+
+    if all_band_names:
+        # Merge common bands
+        projected_rasters = _merge_common_bands(projected_rasters)
+
+        # Merge all bands
+        merge_union = partial(_merge, merge_strategy=MergeStrategy.UNION)  # type: Callable[[_Raster, _Raster], _Raster]
+        raster = reduce(merge_union, projected_rasters)
+
+        return empty.copy_with(image=raster.image, band_names=raster.band_names)
+
+    else:
+        raise ValueError("result contains no bands, use another merge strategy")
+
+
+def _merge_common_bands(rasters):
+    # type: (List[_Raster]) -> List[_Raster]
+    """Combine the common bands.
+
+    """
+    merge_intersect = partial(_merge,
+                              merge_strategy=MergeStrategy.INTERSECTION)  # type: Callable[[_Raster, _Raster], _Raster]
+
+    # Compute band order
+    all_bands = IndexedSet([rs.band_names[0] for rs in rasters])
+
+    def key(rs):
+        return all_bands.index(rs.band_names[0])
+
+    rasters_final = []  # type: List[_Raster]
+    for band_name, rasters_group in groupby(sorted(rasters, key=key), key=key):
+        rasters_final.append(reduce(merge_intersect, rasters_group))
+
+    return rasters_final
+
+
+def _prepare_rasters(rasters, merge_strategy, first):
+    # type: (List[GeoRaster2], MergeStrategy, GeoRaster2) -> Tuple[IndexedSet[str], List[_Raster]]
+    """Prepares the rasters according to the requested ROI and merge strategy.
+
+    """
+    # Create list of prepared rasters
+    all_band_names = IndexedSet(first.band_names)
+    projected_rasters = []  # type: List[GeoRaster2]
+    for raster in rasters:
+        projected_raster = _prepare_other_raster(first, raster)
+
+        # Modify the bands only if an intersecting raster was returned
+        if projected_raster:
+            if merge_strategy is MergeStrategy.INTERSECTION:
+                all_band_names.intersection_update(projected_raster.band_names)
+            elif merge_strategy is MergeStrategy.UNION:
+                all_band_names.update(projected_raster.band_names)
+
+            projected_rasters.append(projected_raster)
+
+    # Extend the rasters list with only those that have the requested bands
+    single_band_rasters = []
+    for projected_raster in projected_rasters:
+        single_band_rasters.extend(_explode_raster(projected_raster, all_band_names))
+
+    return all_band_names, single_band_rasters
+
+
+# noinspection PyDefaultArgument
+def _explode_raster(raster, band_names=[]):
```

Comment:
I think this method could be an object method of GeoRaster, sounds like useful one.

```
g = GeoRaster(..., bands=["red","greed"])
g.explode() => [GeoRaster(band="red"), GeoRaster(band="green")]

```

although in this particular case we might have chosen not to use it because I guess that the implementation will have copy in it, and in current implementation we are not required to do copy, but think about it

Reply:
I agree, and I think the right way to do it is to have a `RasterCollection`. But I preferred to implement that in the next pull request.